### PR TITLE
diags/diag_nvme: call_home command fails on nvmf drive

### PR DESCRIPTION
**Pull request by shroffni**
The diag_nvme command needs to retrive the VPD log page from NVMe for filling in the product data while generating the call-home event. However, call-home feature is supported for directly attached NVMe module. In the current diag_nvme implemenation, if user doesn't provide NVMe device name for diagnostics then it(diag_nvme)  loops through each NVMe moudle (directly connected to the system/LPAR as well as discovered over fabrics) and attempt retrieving the SMART log page as well as VPD page. Unfortunately, diag_nvme fails to retrieve the VPD page for NVMe connected over fabrics and that causes the diag_nvme to print "not-so-nice" failure messages on console.

Henec fixed the diag_nvme code so that for call-home event reporting, it skips the NVMe which is connected over fabrics and prints a "nice-message" informing the user that it's skipping diagnosting for NVMe module connected over fabrics. In a nutshell, with this fix now diag_nvme would only diagnose the NVMe module which is directtly attached (over PCIe) to the system.

### CODE REVIEW

**Comment by sathvika-v** [comment]
Hi @shroffni , Thanks for the PR!

Few comments below..

Checkpatch shows the below warnings:
```
WARNING: Prefer a maximum 75 chars per line (possible unwrapped commit description?)
#7: 
The diag_nvme command needs to retrive the VPD log page from NVMe for filling

WARNING: 'retrive' may be misspelled - perhaps 'retrieve'?
#7: 
The diag_nvme command needs to retrive the VPD log page from NVMe for filling
                               ^^^^^^^

WARNING: 'implemenation' may be misspelled - perhaps 'implementation'?
#10: 
diag_nvme implemenation, if user doesn't provide NVMe device name for
          ^^^^^^^^^^^^^

WARNING: Missing a blank line after declarations
#50: FILE: diags/diag_nvme.c:392:
+		int n = fread(buf, 1, sizeof(buf), fp);
+		if (n) {

total: 0 errors, 4 warnings, 34 lines checked
```

There are a few minor typos in the commit description. Could be changed to `retrieve`, `implementation`, `Hence`, and `directly`.

diag_nvme behavior without this PR:
```
# ./diags/diag_nvme 
Running diagnostics for nvme7
Warn: nvme7 VPD log page ioctl failed: 0x109, fallback to PCIe VPD
/sys/class/nvme/nvme7/device/vpd open failed: No such file or directory
Warn: nvme7 PCIe VPD failed: 0xffffffff, proceed without device VPD data

Running diagnostics for nvme5
Warn: nvme5 VPD log page ioctl failed: 0x109, fallback to PCIe VPD
/sys/class/nvme/nvme5/device/vpd open failed: No such file or directory
Warn: nvme5 PCIe VPD failed: 0xffffffff, proceed without device VPD data

Running diagnostics for nvme3

Running diagnostics for nvme1

Running diagnostics for nvme6
Warn: nvme6 VPD log page ioctl failed: 0x109, fallback to PCIe VPD
/sys/class/nvme/nvme6/device/vpd open failed: No such file or directory
Warn: nvme6 PCIe VPD failed: 0xffffffff, proceed without device VPD data

Running diagnostics for nvme4
Warn: nvme4 VPD log page ioctl failed: 0x109, fallback to PCIe VPD
/sys/class/nvme/nvme4/device/vpd open failed: No such file or directory
Warn: nvme4 PCIe VPD failed: 0xffffffff, proceed without device VPD data

Running diagnostics for nvme2

Running diagnostics for nvme0

NVMe diag command completed successfully
```
diag_nvme behavior with this PR:
```
# ./diags/diag_nvme 
Skipping diagnostics for nvmf : nvme7

Skipping diagnostics for nvmf : nvme5

Running diagnostics for nvme3

Running diagnostics for nvme1

Skipping diagnostics for nvmf : nvme6

Skipping diagnostics for nvmf : nvme4

Running diagnostics for nvme2

Running diagnostics for nvme0

NVMe diag command completed successfully
```

Travis CI builds have passed with this PR.

Except for the minor comments mentioned above, rest of the changes look good to me and the tests have passed as well. 

Tested-by: Sathvika Vasireddy <sv@linux.ibm.com>
Reviewed-by: Sathvika Vasireddy <sv@linux.ibm.com>

Thanks,
Sathvika

**Comment by shroffni** [comment]
Hi Sathvika,

Thank you for testing my patch. I will fix those warnings reported by checkpatch.pl and resubmit the pull-request.

Thanks,
--Nilay

File: diags/diag_nvme.c [hunk 1]
```diff
@@ -375,9 +375,35 @@ static int diagnose_nvme(char *device_name, struct notify *notify, char *file_pa
...
+			 */
+			if (strncmp(buf, "pcie", 4) != 0) {
+				fprintf(stdout, "Skipping diagnostics for nvmf : %s\n",
+						device_name);
+				fclose(fp);
+				return 0;
+			}
+		}
+		fclose(fp);
+	}
```
Comment (maheshsal) [inline diags/diag_nvme.c:410]: How about printing an error message in else case (fp == NULL) to indicate reason for failure to open a file ?

File: diags/diag_nvme.c [hunk 2]
```diff
@@ -375,9 +375,35 @@ static int diagnose_nvme(char *device_name, struct notify *notify, char *file_pa
...
 	unsigned char *raw_data = NULL;
 
+	/*
+	 * Skip diag test if NVMe is connected over fabric
+	 */
+	snprintf(tr_file_path, sizeof(tr_file_path),
+			NVME_SYS_PATH"/%s/%s", device_name, "transport");
+	fp = fopen(tr_file_path, "r");
+	if (fp) {
+		char buf[32];
```
Comment (maheshsal) [inline diags/diag_nvme.c]: It looks like you are only interested in first 4 char 'pcie' from the file. Wouldn't buf[5] is enough ?

**Comment by shroffni** [comment]
On 1/18/24 18:48, Mahesh Salgaonkar wrote:
> @maheshsal commented on this pull request.
> 
> 
> 
>> +		char buf[32];
> +		int n = fread(buf, 1, sizeof(buf), fp);
> +
> +		if (n) {
> +			/*
> +			 * If NVMe transport is anything but pcie then skip the diag test
> +			 */
> +			if (strncmp(buf, "pcie", 4) != 0) {
> +				fprintf(stdout, "Skipping diagnostics for nvmf : %s\n",
> +						device_name);
> +				fclose(fp);
> +				return 0;
> +			}
> +		}
> +		fclose(fp);
> +	}
> 
> How about printing an error message in else case (fp == NULL) to indicate reason for failure to open a file ?
Yeah I would add the code to print error code in this case.

> 
>>   	uint32_t raw_data_len = 0;
>   	unsigned char *raw_data = NULL;
>   
> +	/*
> +	 * Skip diag test if NVMe is connected over fabric
> +	 */
> +	snprintf(tr_file_path, sizeof(tr_file_path),
> +			NVME_SYS_PATH"/%s/%s", device_name, "transport");
> +	fp = fopen(tr_file_path, "r");
> +	if (fp) {
> +		char buf[32];
> 
> It looks like you are only interested in first 4 char 'pcie' from the file. Wouldn't buf[5] is enough ?
> 
That shall be enough for now. For instance, we can have an nvme 
connected over pcie or tcp or rdma or fiber-channel or fabric 
loopback-device and for each of these cases the "transport" file would 
contain "pcie", "tcp", "rdma", "fc" or "loop" respectively. As we 
typically align the size of memory buffer to the power of 2, I would 
prefer the size of 6 bytes. Having said that, IMO, in this case the 
buffer is allocated on the stack and userspcae has large size of stack 
supported, using 32 bytes of buffer space would also not a bad idea 
assuming in future if a new fabric is supported over nvme and kernel 
developer chooses to use a name of transport which exceeds the 6 bytes 
of boundary.

Thanks,
--Nilay

**Comment by shroffni** [comment]
I have updated the patch with following two changes:
- Printing error message when opening "transport" file fails (though 
unlikely) and skip diagnostics in such case
- Set the char buffer size to 12 (buf[12])

Kindly help review changes.

Thanks,
--Nilay

**Pull request merged**